### PR TITLE
Added four different prod servers to run cloud_to_prod(_auth) tests against

### PR DESCRIPTION
**Pull request by adelez**
@jtattermusch

### CODE REVIEW

**Comment by ejona86** [comment]
okay to test

**Comment by ejona86** [comment]
It seems we need another arg other than args.cloud_to_prod since today you can run cloud_to_prod on your desktop, but cloud_to_prod_cloud_gateway can only be run in cloud. I'm not super-familiar with the scripts though. @jtattermusch, does that sound right?

**Comment by ejona86** [comment]
Hmm... I guess it also seems that we will eventually only run cloud_to_prod _or_ cloud_to_prod_cloud_gateway, which is another argument for a separate argument.

**Comment by ejona86** [comment]
Maybe the extra argument can just be "cloud_server_host" or similar and then only run the cloud_to_prod tests once.

(Sorry for the comments back-to-back)

**Comment by jtattermusch** [comment]
Let's go for a more general solution that I'm suggesting in #4922 review.

```
... --cloud_to_prod  --cloud_to_prod_auth   --prod_backend grpc.sandbox.google.com

... --cloud_to_prod  --cloud_to_prod_auth   --prod_backend grpc.sandbox.google.com,grpc.sandbox.googleapis.com 

or even a named version displayed in the report:
... --cloud_to_prod  --cloud_to_prod_auth   --prod_backend regular_gateway=grpc.sandbox.google.com,regular_gateway_v2=grpc.sandbox.google.com, cloud_gateway=123.45.67.89

```

File: tools/run_tests/run_interop_tests.py [hunk 1]
```diff
@@ -422,12 +422,18 @@ def _job_kill_handler(job):
     time.sleep(2)
 
 
-def cloud_to_prod_jobspec(language, test_case, docker_image=None, auth=False):
+def cloud_to_prod_jobspec(language, test_case, use_cloud_gateway=False,
```
Comment (jtattermusch) [inline tools/run_tests/run_interop_tests.py]: If   using cloud gateway only means changing the server_host, we shouldn't  hard code the feature like this.  It is concievable you'll want to add more possible backends in the future and you really don't want to hand-code a boolean parameter for each of these.

**Comment by adelez** [comment]
The third suggestion sounds good to me except that I think prod_backend should be prod_frontend.

**Comment by jtattermusch** [comment]
Another option is to give names to supported prod backends, and then specify list of supported configurations in python code.

```
supported_backends = [
  "default_gateway": ('grpc.sandbox.google.com', 'grpc.sandbox.google.com', ),
  "v2_gateway": ('grpc.sandbox.googleapis.com', 'grpc.sandbox.googleapis.com', errors_allowed=True),
   "cloud_gateway": ('12.13.45.15', 'grpc.sandbox.google.com', errors_allowed=True),
]
```

and then you can provide  list of named backends `--prod_backend=default,v2,cloud_gateway`

**Comment by jtattermusch** [comment]
ad frontend vs backend:   for these tests it's  a prod backend. The reason why you call it a frontend is that you understand the details of how the serving stack works under the hoods (for these test, the details are not really important, it's just the tests are passing against one backend and perhaps not passing against some other backend).
But I don't feel strongly about the naming (provided things makes sense and we stay self-consistent)

**Comment by adelez** [comment]
I updated the code as suggested but how do you plan to input this complicated data structure 
supported_backends = [
  "default_gateway": ('grpc.sandbox.google.com', 'grpc.sandbox.google.com', ),
  "v2_gateway": ('grpc.sandbox.googleapis.com', 'grpc.sandbox.googleapis.com', errors_allowed=True),
   "cloud_gateway": ('12.13.45.15', 'grpc.sandbox.google.com', errors_allowed=True),
]

in command line? I tried with my code it gives me bash: syntax error near unexpected token `('.

**Comment by adelez** [comment]
The idea is by default, it should run against the default gateway and we add the others on Jenkins. But at the same time we allow users to enter any list as they want.

**Comment by jtattermusch** [comment]
I think its fine if from command line the configurations will be referred just by their name (it turns out each configuration has at least three fields, so entering from command line would be to complicated).

Example (naming not really important, just illustrating the idea)
`--prod_server=default_gateway` will be the default 
`--prod_servers=default_gateway,cloud_gateway,...`   will run  all the configurations listed.

File: tools/run_tests/run_interop_tests.py [hunk 2]
```diff
@@ -589,6 +602,11 @@ def aggregate_http2_results(stdout):
                   action='store_const',
                   const=True,
                   help='Run cloud_to_prod_auth tests.')
+argp.add_argument('-p', '--prod_servers',
+                  default=prod_servers,
+                  action='store_const',
+                  const=True,
+                  help='The server to run cloud_to_prod_auth tests against.')
```
Comment (jtattermusch) [inline tools/run_tests/run_interop_tests.py]: nit: not just cloud_to_prod_auth tests, both cloud_to_prod and cloud_to_prod_auth (if requested by corresponding cmdline arguments of course).

**Comment by jtattermusch** [comment]
Did a very quick pass over the good and it seems it's going in a good direction. I will make a more detailed pass tomorrow. Are you planning any more changes?

**Comment by adelez** [comment]
I like the suggestion. Updating the code now. I can easily merge #4922 easily here.

**Comment by adelez** [comment]
Sample command: tools/run_tests/run_interop_tests.py -l c++ --cloud_to_prod -p default_gateway gateway_v2

**Comment by adelez** [comment]
@yang-g

File: tools/run_tests/run_interop_tests.py [hunk 3]
```diff
@@ -589,6 +607,12 @@ def aggregate_http2_results(stdout):
                   action='store_const',
                   const=True,
                   help='Run cloud_to_prod_auth tests.')
+argp.add_argument('-p', '--prod_servers',
```
Comment (jtattermusch) [inline tools/run_tests/run_interop_tests.py]: nit:  short form of the argument "-p" is not very understandable. Can we get rid of the short form and only use --prod_server?

File: tools/run_tests/run_interop_tests.py [hunk 4]
```diff
@@ -572,6 +575,21 @@ def aggregate_http2_results(stdout):
     'percent': 1.0 * passed / (passed + failed)
   }
 
+# A dictionary of prod servers to test. 
+# Format: server_name: (server_host, server_host_override, errors_allowed)
+# TODO(adelez): implement logic for errors_allowed where if the indicated tests
+# fail, they don't impact the overall test result.
+prod_servers = {
+    'default_gateway': ('grpc-test.sandbox.googleapis.com',
```
Comment (jtattermusch) [inline tools/run_tests/run_interop_tests.py]: nit: can we just use "default" instead of "default_gateway" here?

**Comment by jtattermusch** [comment]
LGTM after addressing two minor nits.

As interop test worker is experiencing some trouble, please make sure the tests are passing locally and the HTML report looks good befor merging.

**Comment by jtattermusch** [comment]
I re-ran the interop tests and the HTML report seems to look fine.

File: tools/run_tests/run_interop_tests.py (hunk 3 above)
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: Done

File: tools/run_tests/run_interop_tests.py (hunk 4 above)
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: Done

**Comment by adelez** [comment]
Report looks good.

**Pull request merged**